gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/datafetcher/EntityInteractionsFetcher.java

```diff
@@ -191,6 +275,82 @@ private EntityInteractionQueryResponse executeQueryRequest(
     return new EntityInteractionQueryResponse(entityInteractionQueryRequest, resultSet);
   }
 
+  private List<EntityInteractionQueryRequest> prepareInteractionIdsQueryRequest(
+      RequestContext requestContext,
+      EntitiesRequest request,
+      InteractionsRequest interactionsRequest,
+      boolean incoming) {
+    Set<String> entityTypes = getOtherEntityTypes(interactionsRequest.getFilter());
+    if (entityTypes.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    QueryRequest.Builder builder = QueryRequest.newBuilder();
+    // Filter should include the timestamp filters from parent request first
+    Filter.Builder filterBuilder =
+        Filter.newBuilder()
+            .setOperator(Operator.AND)
+            .addChildFilter(
+                QueryRequestUtil.createBetweenTimesFilter(
+                    AttributeMetadataUtil.getTimestampAttributeId(
+                        metadataProvider, requestContext, SCOPE),
+                    request.getStartTimeMillis(),
+                    request.getEndTimeMillis()));
+
+    this.buildSpaceQueryFilterIfNeeded(requestContext, request.getSpaceId())
+        .ifPresent(filterBuilder::addChildFilter);
+
+    List<String> idColumns =
+        getEntityIdColumnsFromInteraction(
+            DomainEntityType.valueOf(request.getEntityType()), !incoming);
+
+    // Group by the entity id column first, then the other end entity type for the interaction.
+    List<org.hypertrace.core.query.service.api.Expression> idExpressions =
+        idColumns.stream()
+            .map(QueryRequestUtil::createAttributeExpression)
+            .collect(Collectors.toList());
+    builder.addAllGroupBy(idExpressions);
+
+    // adding empty selections
+    List<org.hypertrace.core.query.service.api.Expression> selections = new ArrayList<>();
+    selections.add(
+        QueryRequestUtil.createCountByColumnSelection(
+            Optional.ofNullable(idColumns.get(0)).orElseThrow()));
+
+    QueryRequest protoType = builder.build();
+    Filter protoTypeFilter = filterBuilder.build();
+
+    Map<String, QueryRequest> queryRequests = new HashMap<>();
+
+    // In future we could send these queries in parallel to QueryService so that we can reduce the
+    // response time.
+    for (String e : entityTypes) {
+      DomainEntityType otherEntityType = DomainEntityType.valueOf(e.toUpperCase());
+
+      // Get the filters from the interactions request to 'AND' them with the timestamp filter.
+      Filter.Builder filterCopy = Filter.newBuilder(protoTypeFilter);
+      filterCopy.addChildFilter(
+          convertToQueryFilter(interactionsRequest.getFilter(), otherEntityType));
+
+      QueryRequest.Builder builderCopy = QueryRequest.newBuilder(protoType);
```

Comment:
Similarly, we can avoid building the query request on L320, and directly set filter